Yuhsuan/2144 native wcs channel value

YuHsuan-Hwang opened the pull request:
**Description**

Closes #2144. The issue is caused by the calculation of the channel values in native wcs. There are two kinds of format that defines the pixel size in FITS header: `PCi_j + CDELTi` format and `CDi_j` format, and we only consider `CDELTi` in the calculation.

If the header is in `CDi_j` format, the calculation fails and channel indexes are returned (as in the example of the issue). Frontend crashes if we try to convert the channel indexes to other units.

Another problem in the calculation is that we do not consider `PCi_j`, and the pixel size is a constant. Non-linear spectral axis is not supported (although it's rare in radio images).

In order to solve the problems, obtain the channel values by AST library transformation instead of directly checking the header.

Implementation:

* In `channelInfo`, transform `(xRef, yRef, 0), (xRef, yRef, 1), ...,  (xRef, yRef, channel num - 1)` from pixel coordinate to world coordinate with the 3 dimensional AST FrameSet created from the header entries. `xRef` and `yRef` are constants defined by `CRPIXi` in the header. Origin (0, 0) is used if `CRPIXi` are not found.
* In `getChannelIndexWCS`, convert the wcs value back to pixel coordinate with the same FrameSet, `xRef`, and `yRef`. Adjusted `getTransformedChannel` and `getTransformedChannelList` so that the frameSet remains unchanged.
* Considered rotated cubes when applying transformation.

Code maintenance:

* `delta` is removed from `ChannelInfo` because it's not used.
* `rawValues` is removed from `ChannelInfo` because it's the same with `values`. It was for unit conversion in `ChannelInfo`, but the conversion is moved to other place.

**Checklist**

For linked issues (if there are):
- [x] assignee and label added
- [x] ZenHub issue connection, board status, and estimate updated

For the pull request:
- [x] reviewers and assignee added
- [x] ZenHub estimate, milestone, and release (if needed) added
- [x] e2e test passing / ~corresponding fix added~
- [x] changelog updated / ~no changelog update needed~
- [x] ~protobuf updated to the latest dev commit~ / no protobuf update needed
- [x] `BackendService` unchanged / ~`BackendService` changed and corresponding ICD test fix added~

YuHsuan-Hwang committed:
remove rawValues

YuHsuan-Hwang committed:
adjust indexes

YuHsuan-Hwang committed:
apply ast transformation for channel values in native wcs

YuHsuan-Hwang committed:
remove delta

YuHsuan-Hwang committed:
Merge branch 'dev' into yuhsuan/2144_native_wcs_channel_value

YuHsuan-Hwang committed:
minor change

YuHsuan-Hwang committed:
reformat

veggiesaurus commented on wasm_src/ast_wrapper/ast_wrapper.cc:
```diff
@@ -535,6 +535,29 @@ EMSCRIPTEN_KEEPALIVE int transform3D(AstSpecFrame* wcsinfo, double x, double y,
+        return 1;
+    }
+
+    double in[3][npoint];
```
@YuHsuan-Hwang surely this is not standard C++? We're allocating `3 * npoint` doubles in the stack. Since `npoint` is variable, this shouldn't work :thinking:  I know it's allowed in C99, but not in C++. The correct way to do this would be to allocate the array using malloc on the JS side. Alternatively, you could allocate a single input array of size `npoint * 3` and then just point x, y and z to the appropriate memory locations within?

YuHsuan-Hwang replied:
fixed.

veggiesaurus commented on wasm_src/ast_wrapper/post.ts:
```diff
@@ -197,6 +198,37 @@ Module.transformPointArrays = function (wcsInfo: number, xIn: Float64Array, yIn:
+
+    // Allocate and assign WASM memory
+    const N = xIn.length;
+    const xInPtr = Module._malloc(N * 8);
```
from above, something like

```javascript
const inPtr = Module._malloc(3 * N * 8);
const xInPtr = inPtr;
const yInPtr = xInPtr + 8 * N;
const zInPtr = zInPtr + 8 * N;
```
would assign the right amount of memory and then copy the x, y, and z arrays into it. You could then simply pass `inPtr` in to `transform3DArray` directly

YuHsuan-Hwang replied:
fixed.

YuHsuan-Hwang committed:
fix ast input

YuHsuan-Hwang committed:
Merge branch 'dev' into yuhsuan/2144_native_wcs_channel_value

kswang1029 reviewed (changes requested):
@YuHsuan-Hwang as spotted by e2e tests, if we load a rotated cube, the frequency and velocity info displayed in the cursor info bar in the image viewer is not correct.

> Load Axes-Swapped Cubes                                               | FAIL |
> Element '//*[@id="image-panel-0-0"]/div[4]' should have contained text 'WCS: (9:49:38, 1.420309E+09); Image: (273, 4); Value: -3.61129e-2 Jy/pixel ; Frequency (LSRK): 1420.3000 MHz; Velocity: 22.3201 km/s; Polarization: Stokes Q' but its text was 'WCS: (9:49:38, 1.420309E+09); Image: (273, 4); Value: -3.61129e-2 Jy/pixel ; Frequency (LSRK): -0.0000 MHz; Velocity: 299792.4580 km/s; Polarization: Stokes Q'.

kswang1029 reviewed (changes requested):
@YuHsuan-Hwang as spotted by e2e tests, if we load a rotated cube, the frequency and velocity info displayed in the cursor info bar in the image viewer is not correct.

> Load Axes-Swapped Cubes                                               | FAIL |
> Element '//*[@id="image-panel-0-0"]/div[4]' should have contained text 'WCS: (9:49:38, 1.420309E+09); Image: (273, 4); Value: -3.61129e-2 Jy/pixel ; Frequency (LSRK): 1420.3000 MHz; Velocity: 22.3201 km/s; Polarization: Stokes Q' but its text was 'WCS: (9:49:38, 1.420309E+09); Image: (273, 4); Value: -3.61129e-2 Jy/pixel ; Frequency (LSRK): -0.0000 MHz; Velocity: 299792.4580 km/s; Polarization: Stokes Q'.

kswang1029 commented:
@YuHsuan-Hwang another weird issue spotted also from e2e tests. If we load two identical images (in FITS and HDF5) and create pv image respectively, the result is odd (let me know if you need the same test images).
dev
![check](https://github.com/CARTAvis/carta-frontend/assets/20819712/7c0737b5-331e-4a50-ba7f-82b404b709af)

this branch
![check](https://github.com/CARTAvis/carta-frontend/assets/20819712/67b0ef10-c50b-4de7-8f4d-66cb3a3fb345)

YuHsuan-Hwang committed:
fix get index from native wcs; avoid frameSet being changed when matching

YuHsuan-Hwang committed:
support rotated cube

YuHsuan-Hwang committed:
avoid frameSet being changed when starting animation

YuHsuan-Hwang committed:
Merge branch 'dev' into yuhsuan/2144_native_wcs_channel_value

YuHsuan-Hwang commented:
@kswang1029 
> as spotted by e2e tests, if we load a rotated cube, the frequency and velocity info displayed in the cursor info bar in the image viewer is not correct.

Fixed by considering axis number when applying transformation.
The frequency and velocity info does not work properly when the third axis is shorter when the spectral axis, e.g. 10 < 250 in [432, 250, 10, 1] (RA, FREQ, DEC, STOKES). This exists in the dev branch. We can file a new issue if needed.

> another weird issue spotted also from e2e tests. If we load two identical images (in FITS and HDF5) and create pv image respectively, the result is odd (let me know if you need the same test images).

This also happens after we animate the matched frames. Fixed by avoid the AST FrameSet being changed when matching spectrally and starting animation.

kswang1029 commented:
> @kswang1029
> 
> > as spotted by e2e tests, if we load a rotated cube, the frequency and velocity info displayed in the cursor info bar in the image viewer is not correct.
> 
> Fixed by considering axis number when applying transformation. The frequency and velocity info does not work properly when the third axis is shorter when the spectral axis, e.g. 10 < 250 in [432, 250, 10, 1] (RA, FREQ, DEC, STOKES). This exists in the dev branch. We can file a new issue if needed.
> 
> > another weird issue spotted also from e2e tests. If we load two identical images (in FITS and HDF5) and create pv image respectively, the result is odd (let me know if you need the same test images).
> 
> This also happens after we animate the matched frames. Fixed by avoid the AST FrameSet being changed when matching spectrally and starting animation.

Nice. now there is no regression from e2e tests. I will perform some extra manual tests as this is kinda fundamental change.

kswang1029 commented:
@YuHsuan-Hwang could you have a check to see why the attached cube does not work? 
![Screenshot 2023-06-05 at 15 38 22](https://github.com/CARTAvis/carta-frontend/assets/20819712/0a64b6d1-03aa-4766-8539-a98ae94d34d6)
[ADP.2016-07-12T14_02_12.047_subimage.fits.zip](https://github.com/CARTAvis/carta-frontend/files/11649924/ADP.2016-07-12T14_02_12.047_subimage.fits.zip)

YuHsuan-Hwang committed:
fix free memory

YuHsuan-Hwang commented:
> @YuHsuan-Hwang could you have a check to see why the attached cube does not work?

@kswang1029 Fixed by freeing memory correctly.

kswang1029 reviewed (approved):
I did some cross-comparisons with ds9 and casaviewer using cubes with different spectral conventions such as FREQ, VRAD, AWAV, WAVE, etc. All looking good as far as I can tell. No regression from e2e tests. 

Nice work! 👍

YuHsuan-Hwang committed:
Merge branch 'dev' into yuhsuan/2144_native_wcs_channel_value

YuHsuan-Hwang committed:
update changelog
